src/control/server/storage/bdev/backend_vmd.go

```diff
@@ -0,0 +1,236 @@
+		}
+		vmdAddrs = append(vmdAddrs, strings.TrimSpace(s[0]))
+		i++
+	}
+
+	if len(vmdAddrs) == 0 {
+		return nil, errors.New("error parsing cmd output")
+	}
+
+	return vmdAddrs, nil
+}
+
+// vmdProcessFilters takes an input request and a list of discovered VMD addresses.
+// The VMD addresses are validated against the input request allow and block lists.
+// The output allow list will only contain VMD addresses if either both input allow
+// and block lists are empty or if included in allow and not included in block lists.
+func vmdProcessFilters(inReq *storage.BdevPrepareRequest, vmdPCIAddrs []string) storage.BdevPrepareRequest {
+	var outAllowList []string
+	outReq := *inReq
+
+	if inReq.PCIAllowList == "" && inReq.PCIBlockList == "" {
+		outReq.PCIAllowList = strings.Join(vmdPCIAddrs, storage.BdevPciAddrSep)
+		outReq.PCIBlockList = ""
+		return outReq
+	}
+
+	if inReq.PCIAllowList != "" {
+		allowed := strings.Split(inReq.PCIAllowList, storage.BdevPciAddrSep)
+		for _, addr := range vmdPCIAddrs {
+			if common.Includes(allowed, addr) {
+				outAllowList = append(outAllowList, addr)
+			}
+		}
+		if len(outAllowList) == 0 {
+			// no allowed vmd addresses
+			outReq.PCIAllowList = ""
+			outReq.PCIBlockList = ""
+			return outReq
+		}
+	}
+
+	if inReq.PCIBlockList != "" {
+		var outList []string
+		inList := outAllowList // in case vmdPCIAddrs list has already been filtered
+		if len(inList) == 0 {
+			inList = vmdPCIAddrs
+		}
+		blocked := strings.Split(inReq.PCIBlockList, storage.BdevPciAddrSep)
+		for _, addr := range inList {
+			if !common.Includes(blocked, addr) {
+				outList = append(outList, addr)
+			}
+		}
+		outAllowList = outList
+		if len(outAllowList) == 0 {
+			// no allowed vmd addresses
+			outReq.PCIAllowList = ""
+			outReq.PCIBlockList = ""
+			return outReq
+		}
+	}
+
+	outReq.PCIAllowList = strings.Join(outAllowList, storage.BdevPciAddrSep)
+	outReq.PCIBlockList = ""
+	return outReq
+}
+
+// getVMDPrepReq determines if VMD devices are going to be used and returns a
+// bdev prepare request with the VMD addresses explicitly set in PCI_ALLOWED list.
+//
+// If VMD is not to be prepared, a nil request is returned.
+func getVMDPrepReq(log logging.Logger, req *storage.BdevPrepareRequest, vmdDetect vmdDetectFn) (*storage.BdevPrepareRequest, error) {
+	if !req.EnableVMD {
+		return nil, nil
+	}
+
+	vmdPCIAddrs, err := vmdDetect()
+	if err != nil {
+		return nil, errors.Wrap(err, "VMD could not be enabled")
+	}
+
+	if len(vmdPCIAddrs) == 0 {
+		log.Debug("vmd prep: no vmd devices found")
+		return nil, nil
+	}
+	log.Debugf("volume management devices detected: %v", vmdPCIAddrs)
+
+	vmdReq := vmdProcessFilters(req, vmdPCIAddrs)
+
+	if req.PCIAllowList != "" && vmdReq.PCIAllowList == "" {
+		log.Debugf("vmd prep: %v devices not allowed", vmdPCIAddrs)
+		return nil, nil
+	}
+	if req.PCIBlockList != "" && vmdReq.PCIAllowList == "" {
+		log.Debugf("vmd prep: %v devices blocked", vmdPCIAddrs)
+		return nil, nil
+	}
```

Comment:
I found these conditionals a little hard to understand. Here's a suggestion that could help make it clearer:

```suggestion
	if vmdReq.PCIAllowList == "" {
	        if req.PCIAllowList != "" {
	        	// all VMD devs filtered out by allow list
			log.Debugf("vmd prep: %v devices not allowed", vmdPCIAddrs)
		        return nil, nil
		}
		if req.PCIBlockList != "" {
			// all VMD devs filtered by block list
			log.Debugf("vmd prep: %v devices blocked", vmdPCIAddrs)
			return nil, nil
		}
	}
```